src/pages/fundamentals/common-design-patterns.mdx

```diff
@@ -0,0 +1,453 @@
+libraries and examples to make this easier to handle maintaining these data stuctures. Some data
+structures can be maintained in a single UTxO depending on their contents and structure, others require a
+number of UTxOs and are implemented as linked structures.
+
+Each UTxO contains a reference to the next (and potentially previous) UTxO in the structure, alternatively
+a trie structure could be implemented where each UTxO references any number of children along with a single
+parent.
+
+The contents of a given node on a distributed structure is arbitrary, a key value pair can be held in order
+to create a distributed map using this methodology. Proving exclusion can be done in a single transaction
+as long as there is a cardinal order to the entries, this is done by referencing the two UTxOs which should
+have our entry between them.
+
+## State Thread Tokens (a.k.a STT)
+
+It is often useful to have a mutable state which either changes with each transaction, or on a periodic
+basis. One way to ensure that a datum is not 'spoofed' is to ensure that the input or reference input with
+that datum contains an NFT which has been generated to be unique using one of the method described above.
+
+In this example, we will create an STT that tracks the sum of every transaction that uses the STT.
+
+We will create two different validators:
+
+1. State Thread Token Minting Policy.
+2. Handler of STT.
+
+The STT Minting Policy allows us to create new tokens with a counter datum initialized at 0.
+
+```aiken
+use aiken/collection/dict
+use aiken/collection/list
+use aiken/crypto.{Blake2b_224, Hash, VerificationKey}
+use cardano/assets.{PolicyId, policies}
+use cardano/transaction.{
+  Datum, InlineDatum, OutputReference, Redeemer, Transaction,
+}
+
+validator counter_stt(utxo_ref: OutputReference) {
+  mint(_redeemer: Data, policy_id: PolicyId, self: Transaction) {
+    let Transaction { inputs, outputs, mint, .. } = self
+
+    expect [Pair(_asset_name, quantity)] =
+      mint |> assets.tokens(policy_id) |> dict.to_pairs()
+
+    let is_output_consumed =
+      list.any(inputs, fn(input) { input.output_reference == utxo_ref })
+
+    expect Some(nft_output) =
+      list.find(
+        outputs,
+        fn(output) { list.has(policies(output.value), policy_id) },
+      )
+    expect InlineDatum(datum) = nft_output.datum
+    expect counter: Int = datum
+    is_output_consumed? && (1 == quantity)? && counter == 0
+  }
+}
+```
+Here is the script that will ensure that every transaction increments the value of the counter:
+
+- We check if the transaction is signed by the operator.
+- We check if an input NFT exists and has a datum with an integer.
+- We check if an output exists and has a datum with an integer.
+- We check if the output datum equals the input datum + 1.
+
+```aiken
+type VerificationKeyHash =
+  Hash<Blake2b_224, VerificationKey>
+
+pub type HandlerParams {
+  nft: PolicyId,
+  operator: VerificationKeyHash,
+}
```

Comment:
May I suggest re-writing a bit this example to avoid the params altogether? 

- The `nft` param can be completely skipped by building a multi-validator that takes care of both the `mint` and the `spend` purposes. They would share the same hash, and thus make the parameter redundant. 

- I'd favor using a config here for the operator instead of the applied argument. That is a nice opportunity to showcase the new config capabilities:
 
  **aiken.toml**
  ```toml
  [config.default.operator]
  encoding = "hex"
  bytes = "00000000000000000000000000000000000000000000000000000000"
  ```

  ```
  use config

  // ...
  let is_signed_by_operator =
    list.has(self.extra_signatories, config.operator)
  // ...
  ```

Reply:
Thanks, the feedback has been addressed.